File: 27-matrix-mult/matrix_mult.go
// Golang program to multiply 2 matrices
// Written by M.V.Harish Kumar on 24-05-2024
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
)

func readMatrix(reader *csv.Reader, rows, cols int) [][]int {
	var matrix [][]int
	for i := 0; i < rows; i++ {
		row := make([]int, cols)
		line, _ := reader.Read()
		for j, elem := range line {
			row[j], _ = strconv.Atoi(elem)
		}
		matrix = append(matrix, row)
	}
	return matrix
}

func printMatrix(mat [][]int) {
	for _, row := range mat {
		for _, elem := range row {
			fmt.Printf("%d ", elem)
		}
		fmt.Println()
	}
}

func main() {
	var r1, c1, r2, c2 int
	fmt.Print("Enter dimensions of matrix 1: ")
	fmt.Scanf("%dx%d\n", &r1, &c1)
	fmt.Print("Enter dimensions of matrix 2: ")
	fmt.Scanf("%dx%d\n", &r2, &c2)

	if c1 != r2 {
		fmt.Println("Columns of matrix 1 should be equal to Rows of matrix 2")
		return
	}

	file, _ := os.Open(os.Args[1])
	defer file.Close()
	reader := csv.NewReader(file)
	reader.Comma = ' '

	mat1 := readMatrix(reader, r1, c1)
	mat2 := readMatrix(reader, r2, c2)

	resMat := make([][]int, r1)
	for i := range resMat {
		resMat[i] = make([]int, c2)
	}
	for i := 0; i < r1; i++ {
		for j := 0; j < c2; j++ {
			for k := 0; k < r2; k++ {
				resMat[i][j] += mat1[i][k] * mat2[k][j]
			}
		}
	}

	fmt.Println("Product of:")
	printMatrix(mat1)
	fmt.Println("And")
	printMatrix(mat2)
	fmt.Println("is")
	printMatrix(resMat)
}

File: 27-matrix-mult/matrix_mult.py
# Python program to multiply 2 matrices
# Written by M.V.Harish Kumar on 24-05-2024 
import sys, csv

def read_matrix(file, rows):
    matrix =  [next(file) for _ in range(rows)]
    matrix = list(csv.reader(matrix, delimiter=' '))
    return [[int(x) for x in row] for row in matrix]

def print_matrix(mat):
    for row in mat:
        print(*row)

if __name__ == "__main__":
    r1, c1 = tuple([int(x) for x in input("Enter dimension of matrix 1: ").split('x')])
    r2, c2 = tuple([int(x) for x in input("Enter dimension of matrix 2: ").split('x')])
    if c1 != r2:
        print("Columns of matrix 1 should be equal to Rows of matrix 2")
        exit(-1)
    with open(sys.argv[1]) as f:
        mat1 = read_matrix(f, r1)
        f.readline()
        mat2 = read_matrix(f, r2)
    res_mat = [[sum((a*b for a, b in zip(mat1_row, mat2_col))) for mat2_col in zip(*mat2)] for mat1_row in mat1]
    print("Product of:")
    print_matrix(mat1)
    print("And")
    print_matrix(mat2)
    print("is")
    print_matrix(res_mat)

